### PR TITLE
Migrate program tests from Runtime to Bank

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Testing programs before they are deployed is hard. Before this PR, your options were:

1. Test against KeyedAccounts: They are pure of unit-tests, but are awkward because of the asymmetry. As a user, one creates and sends Transactions to the system, but the program operates on KeyedAccounts.
2. Test against a Bank: They are integration tests, but since the Bank dynamically loads programs, the developer needs to run "cargo build" before "cargo test" to generate the dynamic library in the place where the Bank will look for it.  Forgetting to run "cargo build" results in confusing errors from stale builds.
3. Test against a variation of the Runtime: They are statically-linked integration tests with no sockets and no dynamic loading, but went through a different codepath than the Bank with doesn't verify the transaction's results. You think you've executed an integration test, but actually you bypassed a whole class of Instruction errors.

#### Summary of Changes

* Add new Runtime object that allows injecting instruction processors.
* Expose the new Runtime functionality in the Bank, rendering option 3 obsolete.
* Extend BankClient to process a Script, Instructions, or single Instruction.
* Move Bank-based tests into unit-test suite.

### CODE REVIEW

**Comment by garious** [comment]
Moving this to a WIP. I got it working from Bank. I'll remove all the signing/blockhash noise from BudgetTransaction and move this to BankClient. I'll then delete BankSimulator, because using BankClient makes these tests far more similar to how an DApp would submit them with a thin client.

**Comment by garious** [comment]
@mvines, @CriesofCarrots, can I get your reviews here? This PR effectively renders all `<NAME>Transaction` modules obsolete.

**Comment by garious** [comment]
Thanks @CriesofCarrots. I think the name "script" will make a lot more sense when we abandon the use of the term "program" in favor of "interpreter".

File: programs/config/src/lib.rs [hunk 1]
```diff
@@ -78,125 +79,97 @@ mod tests {
...
+        let script = Script::new(vec![instruction]);
+        let mut transaction = script.compile();
         transaction.instructions[0].data = vec![0; 123];
-
-        let mut accounts = vec![config_account];
-        process_transaction(&transaction, &mut accounts).unwrap_err();
-    }
-
-    #[test]
-    fn test_process_store_fail_account0_invalid_owner() {
```
Comment (mvines) [inline programs/config/src/lib.rs]: We can’t represent this test in the new model?

Comment (garious) [reply programs/config/src/lib.rs]: Done

**Pull request merged**